sympy/algebras/quaternion.py

```diff
@@ -0,0 +1,545 @@
+# References :
+# http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/
+# https://en.wikipedia.org/wiki/Quaternion
+
+from __future__ import print_function
+
+from sympy.core.expr import Expr
+from sympy import Rational
+from sympy import re, im, conjugate
+from sympy import sqrt, sin, cos, acos, asin, exp, ln
+from sympy import trigsimp
+from sympy import diff, integrate
+from sympy import Matrix, Add, Mul
+from sympy import symbols, sympify
+from sympy.printing.latex import latex
+from sympy.printing import StrPrinter
+
+
+class Quaternion(Expr):
+    """Provides basic quaternion operations.
+    Quaternion objects can be instantiated as Quaternion(a, b, c, d)
+    as in (a + b*i + c*j + d*k).
+
+    Example
+    ========
+
+    >>> from sympy.algebras.quaternion import Quaternion
+    >>> q = Quaternion(1, 2, 3, 4)
+    >>> q
+    1 + 2*i + 3*j + 4*k
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, a=0, b=0, c=0, d=0, real_elements = True):
```

Comment:
Please document the real_elements flag. 

From what I understand, it differentiates whether `i`, the quaternion, is treated as sqrt(-1). I think a better name could be used. How about `i_is_complex=False`.

Reply:
I was just asking if this is a good way to achieve add and mul in quaternions over the real numbers and quaternions over the complex numbers.

Reply:
what about `field=S.Reals` instead?

Reply:
@Upabjojr I think that would be a better name. Should I let the values if the elements influence this flag or should it be left to the user to change it?